---
title: "04 promoters / glioblastoma"
params:
  seed: 1
author: "Izaskun Mallona"
output:
  html_document:
    keep_md: false
    toc: true
    toc_float: true
    toc_depth: 4

---

# Notes and todos

From https://www.nature.com/articles/s41591-018-0156-x#Sec10

```
The epiallele entropy (EPY), a measure of subclonality within a tumor, was calculated using a slightly modified version of methclone (v0.1)50. We calculated epiallele entropies separately for each of the samples and, independently, for each matched pair of primary and recurring tumors. Input files to methclone were created by aligning the trimmed RRBS reads to the human reference genome (GRCh38) using Bismark58. To find suitable thresholds for read coverage and entropy change, we performed a series of analyses with read thresholds set to 20, 40, and 60 reads and entropy change thresholds set to –40, –50, –60, –70, –80, and –90. We did not observe significant differences in the results except for the expected effects at the extremes of the spectrum. Therefore, we chose a moderate read threshold of 40 reads for methclone to consider a locus, and loci with a combinatorial entropy change below –80 were classified as epigenetic shift loci (eloci) between primary and recurring tumors or normal brain control samples (as described in the original publication)50. For each pair, we then calculated epiallele shifts per million loci (EPM), dividing the number of eloci by the total number of assessed loci normalized to 1 million loci50.

Sample-wise PDR and epiallele entropy values were calculated by averaging across all promoters that were covered in >75% of the samples. Promoter regions were defined as the genomic region 1 kb upstream to 500 bp downstream of a given transcription start site as annotated by GENCODE72.

```

But look at figure 5d: is clear it does not allow to stratify patients in this case! https://www.nature.com/articles/s41591-018-0156-x/figures/5

Plus match patients, recurrence etc


# Configure

```{r}
NTHREADS = 5

DATA_PATH= '/home/imallona/src/meth_entropies/runs/glios'
CLINICAL_DATA_PATH = file.path(DATA_PATH,  'data', '41591_2018_156_MOESM3_ESM.csv')
```


```{r knitr_setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.width = 6,
                      fig.height = 6,
                      comment = NULL,
                      warning = TRUE,
                      cache = TRUE, tidy = TRUE, tidy.opts = list(width.cutoff=80))

## for transparencies
## setHook(packageEvent("grDevices", "onLoad"),
## function(...) grDevices::X11.options(type='cairo'))
## options(device='x11')
options(bitmapType="cairo")

```



```{r tag}
analysis_tag <- '04_promoters_entropy_survival'
```

```{r}
library(devtools)
library(survival)
library(maxstat)
library(entropy)
library('survminer')


```

# Data load

Methylation and entropy data (averaged by non-overlapping promoter)

```{r}
d <- list(meth = list(),
          entropy = list())

for (fn in list.files(DATA_PATH,
                      pattern = '.*average_entropy_promoter.bed.gz', recursive = TRUE)) {
    tryCatch({
        d$entropy[[dirname(fn)]] <- read.table(file.path(DATA_PATH, fn),
                                               header = FALSE, stringsAsFactors = FALSE,
                                       col.names = c('promoter', 'avg_entropy'))
    }, error = function(x) print(fn))
    
}

for (fn in list.files(DATA_PATH,
                      pattern = '.*average_methylation_promoter.bed.gz', recursive = TRUE)) {
    tryCatch({
        d$meth[[dirname(fn)]] <- read.table(file.path(DATA_PATH, fn),
                                            header = FALSE, stringsAsFactors = FALSE,
                                       col.names = c('promoter', 'avg_meth'))
    }, error = function(x) print(fn))
    
}

```

Clinical data from https://www.nature.com/articles/s41591-018-0156-x, e.g. https://static-content.springer.com/esm/art%3A10.1038%2Fs41591-018-0156-x/MediaObjects/41591_2018_156_MOESM3_ESM.xlsx


```{r}
clin <- read.table(CLINICAL_DATA_PATH, header = TRUE, sep = '\t',
                   stringsAsFactors = FALSE,
                   na.strings = "N/A")

colnames(clin)[1:2] <- c('patient',  'sample')
clin$id <- sprintf('%s_%s', clin$patient, clin$sample)

```

Building a per-patient and sample entropies/methylation dataframe.

```{r}

dict <- data.frame(raw_id = names(d$entropy),
                   entropy = NA,
                   meth = NA,
                   id = NA,
                   stringsAsFactors = FALSE)

dict$clin_id <- gsub('_NA', '',
                     sapply(strsplit(split = "_", dict$raw_id),
                            function(x) return(sprintf('%s_%s_%s_%s_%s',
                                                  x[2], x[3], x[4], x[5], x[6]))))

dict$patient <- sapply(strsplit(split = "_", dict$raw_id),
                       function(x) return(sprintf('%s_%s', x[2], x[3])))

dict$sample <- gsub('_NA', '', sapply(strsplit(split = "_", dict$raw_id),
                                      function(x) return(sprintf('%s_%s_%s', x[4], x[5], x[6]))))


dict$id <- sprintf('%s_%s', dict$patient, dict$sample)
dict$entropy <- NA

rownames(dict) <- dict$raw_id

```

Select promoters present in at least 75% of the patients


```{r}

## promoters <- as.data.frame(table(unlist(sapply(d$entropy, function(x) return(x$promoter)))))

## ## promoters$percentage <- promoters$Freq/length(d$entropy)

## tmp <- lapply(d$entropy, function(x) merge(x = data.frame(promoter = promoters),
##                                            y = x,
##                                            by = 'promoter',
##                                            all.x = TRUE))


promoters <- as.data.frame(table(unlist(lapply(d$entropy,
                                               function(x) return(unique(x$promoter))))))

promoters$percentage <- promoters$Freq/length(d$entropy)

## present in 75% of the samples at least
selected <- promoters[promoters$percentage >= 0.75, 'Var1']

## we'll remove readouts from different promoters as well
selected <- grep(',', selected, invert = TRUE, value = TRUE)

```

```{r}
length(selected)
```

Filtering out data represented in less than 75% of the samples.


```{r}

d$entropy <- lapply(d$entropy, function(x) x[x$promoter %in% selected,])
d$meth <- lapply(d$meth, function(x) x[x$promoter %in% selected,])
```

Remove samples with no data after filtering


```{r}

d$meth <- d$meth[lengths(d$meth) != 0]
d$entropy <- d$entropy[lengths(d$entropy) != 0]
```

Standardization


```{r, eval = FALSE}
print('does not work on averages')

standardize_without_coverage <- function(id, meth, entropy) {
    cov = 10
    
    d <- data.frame(id = id, meth = meth, entropy = entropy)
    ## rownames(d) <- d$id
    
    d$pU <- (rep(cov, length(meth)) - (cov * meth))/cov
    d$pM <- 1 - d$pU
    
    d$max_entropy <- apply(data.frame(pUU = d$pU^2,
                                      pUM = d$pU * d$pM,
                                      pMU = d$pU * d$pM,
                                      pMM = d$pM^2),
                           1, entropy)


    d$min_entropy_full <- apply(data.frame(pUU = d$pU,
                                           pUM = 0,
                                           pMU = 0,
                                           pMM = d$pM),
                                1, entropy)

    
    
    d$min_entropy_meth <- apply(data.frame(pUU = 0,
                                           pUM = abs(d$pM - d$pU)/2 ,
                                           pMU =  0,
                                           pMM = d$pM),
                                1, entropy)

    d$min_entropy_meth[d$meth >= 0.5] <- apply(
        data.frame(pU = 0,
                   pUM = abs(d[meth >= 0.5,]$pM - d[meth >= 0.5,]$pU)/2 ,
                   pMU =  0,
                   pMM = d[meth >= 0.5,]$pU),
        1, entropy)
    
    
    d$lower_bound <- apply(d[,c('min_entropy_full', 'min_entropy_meth')], 1, min)

    
    d$standardized_entropy <- (d$entropy - d$lower_bound) / (d$max_entropy - d$lower_bound)
    d$standardized_entropy[is.na(d$standardized_entropy)] <- 0

    return(d[, c('id', 'standardized_entropy')])
}


d$standardized <- list()
for (item in names(d$entropy)) {
    tryCatch({
        d$standardized[[item]] <- standardize_without_coverage(
            id = d$entropy[[item]]$promoter,
            meth = d$meth[[item]]$avg_meth,
            entropy = d$entropy[[item]]$avg_entropy)
    }, error = function(x) print(sprintf('%s %s', item, x)))
}
```


Get the mean value. By patient and same type (e.g. whether fv or not; not by patient)


```{r}
for (id in dict$raw_id) {
    dict[id, 'meth'] <- mean(d$meth[[id]]$avg_meth, na.rm = TRUE)
    
    dict[id, 'entropy'] <- mean(d$entropy[[id]]$avg_entropy, na.rm = TRUE)

    ## dict[id, 'standardized_entropy'] <- mean(d$standardized[[id]]$standardized_entropy,
    ##                                          na.rm = TRUE)
}

```

# Survival

BEWARE WE ARE TAKING DOUBLE THE CASES WE HAVE, BECAUSE OF THE PROCESSING OF FV AND NONFV: split them not to falsely increase the cohort size!

## Global survival all


Survival object update.




```{r}
dict$another_id <- gsub('_fv', '', dict$clin_id)

tmp <- list(fv = dict[grep('fv', dict$id),],
            nonfv =  dict[grep('fv', dict$id, invert = TRUE),])

dict <- tmp

## commenting out because the identifier (same patient <-> fv and not fv data)
## dict$fv <-  merge(dict$fv, clin, by.y = 'id', by.x = 'clin_id')

dict$fv <-  merge(dict$fv, clin, by.y = 'id', by.x = 'another_id')

dict$nonfv <-  merge(dict$nonfv, clin, by.y = 'id', by.x = 'clin_id')

dict$fv$fv <- 'fv'
dict$nonfv$fv <- 'nonfv' 
dict <- do.call(rbind.data.frame, dict)

```

```{r}

dict$event_global <- NA
dict$event_global[dict$Status == 'dead'] <- 1
dict$event_global[dict$Status == 'alive'] <- 0
```

Global survival vs methylation

```{r}
tag <- 'avg_meth'
tag2 <- tag

ctime <- summary(coxph(Surv(dict$Follow.up.Time..months., dict$event_global) ~
                           dict$meth))

## ctime_cutoff <- maxstat.test(Surv(Follow.up.Time..months., event_global) ~
##                                  dict$meth,
##                              data = dict,
##                              smethod = "LogRank", pmethod = "condMC", B = 9999,
##                              alpha = 0.05)

ctime_cutoff <- list(estimate = median(na.omit(dict$meth)))

fit <- survfit(Surv(dict$Follow.up.Time..months., dict$event_global)
               ~ strata(dict$meth < ctime_cutoff$estimate))

ctime_sdf <- survdiff(Surv(dict$Follow.up.Time..months., dict$event_global)
                      ~ dict$meth < ctime_cutoff$estimate , rho = 0)

ctime_sdf$p.value <- 1 - pchisq(ctime_sdf$chisq, length(ctime_sdf$n) - 1)
ctime_sdf$statistic <-ctime_sdf$chisq


plot(fit, lty = 1:2,
     ylab = 'event global survival probability',
     xlab = 'time global',
     main = sprintf('%s\ncoxph pvalue %s statistic %s\nlogrank pvalue %s statistic %s',
                    tag, tag2,
                    sprintf('%g', coef(ctime)[5]), sprintf('%g', coef(ctime)[1]),
                    sprintf('%g', ctime_sdf$p.value),  sprintf('%g', ctime_sdf$statistic)))

legend(
    "bottomleft",
    legend = c(sprintf('x < %s', ctime_cutoff$estimate),
               sprintf('x >= %s', ctime_cutoff$estimate)),
    lty = 1:2,
    horiz = FALSE,
    bty = 'n')

## plot(ctime_cutoff, xlab = tag, main = 'cut-off selection global survival')
    
```


Global survival vs entropy

```{r}
tag <- 'avg_entropy'
tag2 <- tag

ctime <- summary(coxph(Surv(dict$Follow.up.Time..months., dict$event_global) ~
                           dict$entropy))

## ctime_cutoff <- maxstat.test(Surv(Follow.up.Time..months., event_global) ~
##                                  entropy,
##                              data = dict,
##                              smethod = "LogRank", pentropyod = "condMC", B = 9999,
##                              alpha = 0.05)

ctime_cutoff <- list(estimate = median(na.omit(dict$entropy)))

fit <- survfit(Surv(dict$Follow.up.Time..months., dict$event_global)
               ~ strata(dict$entropy < ctime_cutoff$estimate))

ctime_sdf <- survdiff(Surv(dict$Follow.up.Time..months., dict$event_global)
                      ~ dict$entropy < ctime_cutoff$estimate , rho = 0)

ctime_sdf$p.value <- 1 - pchisq(ctime_sdf$chisq, length(ctime_sdf$n) - 1)
ctime_sdf$statistic <-ctime_sdf$chisq


plot(fit, lty = 1:2,
     ylab = 'event global survival probability',
     xlab = 'time global',
     main = sprintf('%s\ncoxph pvalue %s statistic %s\nlogrank pvalue %s statistic %s',
                    tag, tag2,
                    sprintf('%g', coef(ctime)[5]), sprintf('%g', coef(ctime)[1]),
                    sprintf('%g', ctime_sdf$p.value),  sprintf('%g', ctime_sdf$statistic)))

legend(
    "bottomleft",
    legend = c(sprintf('x < %s', ctime_cutoff$estimate),
               sprintf('x >= %s', ctime_cutoff$estimate)),
    lty = 1:2,
    horiz = FALSE,
    bty = 'n')

## plot(ctime_cutoff, xlab = tag, main = 'cut-off selection global survival')
    
```

Prettier output
```{r}

library('survminer')
library(survival)

tag <- 'avg_entropy'
tag2 <- tag

ctime <- summary(coxph(Surv(dict$Follow.up.Time..months., dict$event_global) ~
                           dict$entropy))

## ctime_cutoff <- maxstat.test(Surv(Follow.up.Time..months., event_global) ~
##                                  entropy,
##                              data = dict,
##                              smethod = "LogRank", pentropyod = "condMC", B = 9999,
##                              alpha = 0.05)

ctime_cutoff <- list(estimate = median(na.omit(dict$entropy)))

fit <- survfit(Surv(dict$Follow.up.Time..months., dict$event_global)
               ~ strata(dict$entropy < ctime_cutoff$estimate))

ctime_sdf <- survdiff(Surv(dict$Follow.up.Time..months., dict$event_global)
                      ~ dict$entropy < ctime_cutoff$estimate , rho = 0)

ctime_sdf$p.value <- 1 - pchisq(ctime_sdf$chisq, length(ctime_sdf$n) - 1)
ctime_sdf$statistic <-ctime_sdf$chisq


tmp <- data.frame(months = dict$Follow.up.Time..months.,
                  event_global = dict$event_global,
                  strata = dict$entropy < ctime_cutoff$estimate)

fit <- survfit(Surv(tmp$months, tmp$event_global)
               ~ tmp$strata)


psurv <- ggsurvplot(
    fit, 
    data = tmp, 
    size = 1,                 # change line size
    palette = 
        c("#E7B800", "#2E9FDF"),# custom color palettes
    conf.int = TRUE,          # Add confidence interval
    pval = TRUE,              # Add p-value
  risk.table = TRUE,        # Add risk table
  risk.table.col = "strata",# Risk table color by groups
  legend.labs = 
      c("Low entropy", "High entropy"),    # Change legend labels
  risk.table.height = 0.25, # Useful to change when you have multiple groups
  ggtheme = theme_bw()      # Change ggplot2 theme
)

print(psurv)

ggsave(filename = 'psurv_entropy.png', plot = print(psurv), width = 5, height = 5)


```

Complex models

No fv

```{r}

tag <- 'avg_entropy'
tag2 <- tag

(ctime1 <- summary(coxph(Surv(dict$Follow.up.Time..months., dict$event_global) ~
                            dict$meth)))

(ctime2 <- summary(coxph(Surv(dict$Follow.up.Time..months., dict$event_global) ~
                            dict$entropy +
                            dict$meth +
                            dict$Age.at.Diagnosis +
                            dict$Sex +
                            dict$IDH.status +
                            dict$fv)))


```


## Free survival all

Disease-free survival
Disease recidives: some are NA, let's encode them as 0

```{r}
dict$event_free <- NA

dict$event_free[dict$Time.to.Progression..months. < dict$Follow.up.Time..months.] <- 1
dict$event_free[dict$Time.to.Progression..months. > dict$Follow.up.Time..months.] <- 0
dict$event_free[is.na(dict$event_free)] <- 0

table(dict$event_free)
```

```{r}

tag <- 'avg_meth'
tag2 <- 'free survival'

ctime <- summary(coxph(Surv(dict$Follow.up.Time..months., dict$event_free) ~
                           dict$meth))

## ctime_cutoff <- maxstat.test(Surv(Follow.up.Time..months., event_global) ~
##                                  dict$meth,
##                              data = dict,
##                              smethod = "LogRank", pmethod = "condMC", B = 9999,
##                              alpha = 0.05)

ctime_cutoff <- list(estimate = median(na.omit(dict$meth)))

fit <- survfit(Surv(dict$Follow.up.Time..months., dict$event_free)
               ~ strata(dict$meth < ctime_cutoff$estimate))

ctime_sdf <- survdiff(Surv(dict$Follow.up.Time..months., dict$event_global)
                      ~ dict$meth < ctime_cutoff$estimate , rho = 0)

ctime_sdf$p.value <- 1 - pchisq(ctime_sdf$chisq, length(ctime_sdf$n) - 1)
ctime_sdf$statistic <-ctime_sdf$chisq


plot(fit, lty = 1:2,
     ylab = 'event free survival probability',
     xlab = 'time global',
     main = sprintf('%s\ncoxph pvalue %s statistic %s\nlogrank pvalue %s statistic %s',
                    tag, tag2,
                    sprintf('%g', coef(ctime)[5]), sprintf('%g', coef(ctime)[1]),
                    sprintf('%g', ctime_sdf$p.value),  sprintf('%g', ctime_sdf$statistic)))

legend(
    "bottomleft",
    legend = c(sprintf('x < %s', ctime_cutoff$estimate),
               sprintf('x >= %s', ctime_cutoff$estimate)),
    lty = 1:2,
    horiz = FALSE,
    bty = 'n')

## plot(ctime_cutoff, xlab = tag, main = 'cut-off selection global survival')


```



```{r}

tag <- 'avg_entropy'
tag2 <- 'free survival'

ctime <- summary(coxph(Surv(dict$Follow.up.Time..months., dict$event_free) ~
                           dict$entropy))

## ctime_cutoff <- maxstat.test(Surv(Follow.up.Time..months., event_global) ~
##                                  dict$entropy,
##                              data = dict,
##                              sentropyod = "LogRank", pentropyod = "condMC", B = 9999,
##                              alpha = 0.05)

ctime_cutoff <- list(estimate = median(na.omit(dict$entropy)))

fit <- survfit(Surv(dict$Follow.up.Time..months., dict$event_free)
               ~ strata(dict$entropy < ctime_cutoff$estimate))

ctime_sdf <- survdiff(Surv(dict$Follow.up.Time..months., dict$event_global)
                      ~ dict$entropy < ctime_cutoff$estimate , rho = 0)

ctime_sdf$p.value <- 1 - pchisq(ctime_sdf$chisq, length(ctime_sdf$n) - 1)
ctime_sdf$statistic <-ctime_sdf$chisq


plot(fit, lty = 1:2,
     ylab = 'event free survival probability',
     xlab = 'time global',
     main = sprintf('%s\ncoxph pvalue %s statistic %s\nlogrank pvalue %s statistic %s',
                    tag, tag2,
                    sprintf('%g', coef(ctime)[5]), sprintf('%g', coef(ctime)[1]),
                    sprintf('%g', ctime_sdf$p.value),  sprintf('%g', ctime_sdf$statistic)))

legend(
    "bottomleft",
    legend = c(sprintf('x < %s', ctime_cutoff$estimate),
               sprintf('x >= %s', ctime_cutoff$estimate)),
    lty = 1:2,
    horiz = FALSE,
    bty = 'n')

## plot(ctime_cutoff, xlab = tag, main = 'cut-off selection global survival')


```

## Global survival FV meth

```{r}
tag <- 'avg_meth FV only'
tag2 <- tag

dictfv<- dict[dict$fv =='fv',]

ctime <- summary(coxph(Surv(dictfv$Follow.up.Time..months., dictfv$event_global) ~
                           dictfv$meth))

## ctime_cutoff <- maxstat.test(Surv(Follow.up.Time..months., event_global) ~
##                                  dictfv$meth,
##                              data = dictfv,
##                              smethod = "LogRank", pmethod = "condMC", B = 9999,
##                              alpha = 0.05)

ctime_cutoff <- list(estimate = median(na.omit(dictfv$meth)))

fit <- survfit(Surv(dictfv$Follow.up.Time..months., dictfv$event_global)
               ~ strata(dictfv$meth < ctime_cutoff$estimate))

ctime_sdf <- survdiff(Surv(dictfv$Follow.up.Time..months., dictfv$event_global)
                      ~ dictfv$meth < ctime_cutoff$estimate , rho = 0)

ctime_sdf$p.value <- 1 - pchisq(ctime_sdf$chisq, length(ctime_sdf$n) - 1)
ctime_sdf$statistic <-ctime_sdf$chisq


plot(fit, lty = 1:2,
     ylab = 'event global survival probability',
     xlab = 'time global',
     main = sprintf('%s\ncoxph pvalue %s statistic %s\nlogrank pvalue %s statistic %s',
                    tag, tag2,
                    sprintf('%g', coef(ctime)[5]), sprintf('%g', coef(ctime)[1]),
                    sprintf('%g', ctime_sdf$p.value),  sprintf('%g', ctime_sdf$statistic)))

legend(
    "bottomleft",
    legend = c(sprintf('x < %s', ctime_cutoff$estimate),
               sprintf('x >= %s', ctime_cutoff$estimate)),
    lty = 1:2,
    horiz = FALSE,
    bty = 'n')

## plot(ctime_cutoff, xlab = tag, main = 'cut-off selection global survival')
    
```


## Global survival FV entropy

```{r}
tag <- 'avg_entropy FV only'
tag2 <- tag

dictfv<- dict[dict$fv =='fv',]

ctime <- summary(coxph(Surv(dictfv$Follow.up.Time..months., dictfv$event_global) ~
                           dictfv$entropy))

## ctime_cutoff <- maxstat.test(Surv(Follow.up.Time..months., event_global) ~
##                                  dictfv$entropy,
##                              data = dictfv,
##                              sentropyod = "LogRank", pentropyod = "condMC", B = 9999,
##                              alpha = 0.05)

ctime_cutoff <- list(estimate = median(na.omit(dictfv$entropy)))

fit <- survfit(Surv(dictfv$Follow.up.Time..months., dictfv$event_global)
               ~ strata(dictfv$entropy < ctime_cutoff$estimate))

ctime_sdf <- survdiff(Surv(dictfv$Follow.up.Time..months., dictfv$event_global)
                      ~ dictfv$entropy < ctime_cutoff$estimate , rho = 0)

ctime_sdf$p.value <- 1 - pchisq(ctime_sdf$chisq, length(ctime_sdf$n) - 1)
ctime_sdf$statistic <-ctime_sdf$chisq


plot(fit, lty = 1:2,
     ylab = 'event global survival probability',
     xlab = 'time global',
     main = sprintf('%s\ncoxph pvalue %s statistic %s\nlogrank pvalue %s statistic %s',
                    tag, tag2,
                    sprintf('%g', coef(ctime)[5]), sprintf('%g', coef(ctime)[1]),
                    sprintf('%g', ctime_sdf$p.value),  sprintf('%g', ctime_sdf$statistic)))

legend(
    "bottomleft",
    legend = c(sprintf('x < %s', ctime_cutoff$estimate),
               sprintf('x >= %s', ctime_cutoff$estimate)),
    lty = 1:2,
    horiz = FALSE,
    bty = 'n')

## plot(ctime_cutoff, xlab = tag, main = 'cut-off selection global survival')
    
```




## Global survival non FV meth

```{r}
tag <- 'avg_meth nonFV only'
tag2 <- tag

dictnonfv<- dict[dict$fv =='nonfv',]

ctime <- summary(coxph(Surv(dictnonfv$Follow.up.Time..months., dictnonfv$event_global) ~
                           dictnonfv$meth))

## ctime_cutoff <- maxstat.test(Surv(Follow.up.Time..months., event_global) ~
##                                  dictnonfv$meth,
##                              data = dictnonfv,
##                              smethod = "LogRank", pmethod = "condMC", B = 9999,
##                              alpha = 0.05)

ctime_cutoff <- list(estimate = median(na.omit(dictnonfv$meth)))

fit <- survfit(Surv(dictnonfv$Follow.up.Time..months., dictnonfv$event_global)
               ~ strata(dictnonfv$meth < ctime_cutoff$estimate))

ctime_sdf <- survdiff(Surv(dictnonfv$Follow.up.Time..months., dictnonfv$event_global)
                      ~ dictnonfv$meth < ctime_cutoff$estimate , rho = 0)

ctime_sdf$p.value <- 1 - pchisq(ctime_sdf$chisq, length(ctime_sdf$n) - 1)
ctime_sdf$statistic <-ctime_sdf$chisq


plot(fit, lty = 1:2,
     ylab = 'event global survival probability',
     xlab = 'time global',
     main = sprintf('%s\ncoxph pvalue %s statistic %s\nlogrank pvalue %s statistic %s',
                    tag, tag2,
                    sprintf('%g', coef(ctime)[5]), sprintf('%g', coef(ctime)[1]),
                    sprintf('%g', ctime_sdf$p.value),  sprintf('%g', ctime_sdf$statistic)))

legend(
    "bottomleft",
    legend = c(sprintf('x < %s', ctime_cutoff$estimate),
               sprintf('x >= %s', ctime_cutoff$estimate)),
    lty = 1:2,
    horiz = FALSE,
    bty = 'n')

## plot(ctime_cutoff, xlab = tag, main = 'cut-off selection global survival')
    
```



## Global survival non FV entropy

```{r}
tag <- 'avg_entropy nonFV only'
tag2 <- tag

dictnonfv<- dict[dict$fv =='nonfv',]

ctime <- summary(coxph(Surv(dictnonfv$Follow.up.Time..months., dictnonfv$event_global) ~
                           dictnonfv$entropy))

## ctime_cutoff <- maxstat.test(Surv(Follow.up.Time..months., event_global) ~
##                                  dictnonfv$entropy,
##                              data = dictnonfv,
##                              sentropyod = "LogRank", pentropyod = "condMC", B = 9999,
##                              alpha = 0.05)

ctime_cutoff <- list(estimate = median(na.omit(dictnonfv$entropy)))

fit <- survfit(Surv(dictnonfv$Follow.up.Time..months., dictnonfv$event_global)
               ~ strata(dictnonfv$entropy < ctime_cutoff$estimate))

ctime_sdf <- survdiff(Surv(dictnonfv$Follow.up.Time..months., dictnonfv$event_global)
                      ~ dictnonfv$entropy < ctime_cutoff$estimate , rho = 0)

ctime_sdf$p.value <- 1 - pchisq(ctime_sdf$chisq, length(ctime_sdf$n) - 1)
ctime_sdf$statistic <-ctime_sdf$chisq


plot(fit, lty = 1:2,
     ylab = 'event global survival probability',
     xlab = 'time global',
     main = sprintf('%s\ncoxph pvalue %s statistic %s\nlogrank pvalue %s statistic %s',
                    tag, tag2,
                    sprintf('%g', coef(ctime)[5]), sprintf('%g', coef(ctime)[1]),
                    sprintf('%g', ctime_sdf$p.value),  sprintf('%g', ctime_sdf$statistic)))

legend(
    "bottomleft",
    legend = c(sprintf('x < %s', ctime_cutoff$estimate),
               sprintf('x >= %s', ctime_cutoff$estimate)),
    lty = 1:2,
    horiz = FALSE,
    bty = 'n')

## plot(ctime_cutoff, xlab = tag, main = 'cut-off selection global survival')
    
```


## Chunks



Standardized entropy?


```{r}

```


What about the proportion of highly entropic regions?

```{r untested_m, eval = FALSE}

dict <- data.frame(entropy = names(d$entropy),
                   meth = NA,
                   id = NA,
                   stringsAsFactors = FALSE)

dict$patient <- sapply(strsplit(split = "_", dict$entropy),
                       function(x) return(sprintf('%s_%s', x[2], x[3])))

dict$sample <- gsub('_NA', '', sapply(strsplit(split = "_", dict$entropy),
                                      function(x) return(sprintf('%s_%s_%s', x[4], x[5], x[6]))))

dict$id <- sprintf('%s_%s', dict$patient, dict$sample)
dict$entropy <- NA

rownames(dict) <- dict$id

dict <- merge(dict, clin, by = 'id')

dict$event_global <- NA
dict$event_global[dict$Status == 'dead'] <- 1
dict$event_global[dict$Status == 'alive'] <- 0



dict$count_entropic <- NA
for (id in dict$id) {
    entropy_id <- grep(pattern = sprintf('.*%s$', id), x = names(d$entropy), value = TRUE)
    if (length(entropy_id) == 1) {
        ## print(entropy_id)
        ## tryCatch({
        dict[entropy_id, 'count_entropic'] <- sum(d$entropy[[entropy_id]]$avg_entropy > 0.5,
                                                  na.rm = TRUE)/length(d$entropy[[entropy_id]]$avg_entropy)

        
        ## }, error = function(x) print(''))
    }
}

summary(dict$count_entropic)
summary(dict$event_global)


tag <- 'avg_count_entropic'
tag2 <- tag

ctime <- summary(coxph(Surv(dict$Follow.up.Time..months., dict$event_global) ~
                           dict$Sex))

ctime <- summary(coxph(Surv(dict$Follow.up.Time..months., dict$event_global) ~
                           dict$count_entropic))

ctime_cutoff <- maxstat.test(Surv(Follow.up.Time..months., event_global) ~
                                 count_entropic,
                             data = dict,
                             smethod = "LogRank", pcount_entropicod = "condMC", B = 9999,
                             alpha = 0.05)

fit <- survfit(Surv(dict$Follow.up.Time..months., dict$event_global)
               ~ strata(dict$count_entropic < ctime_cutoff$estimate))

ctime_sdf <- survdiff(Surv(dict$Follow.up.Time..months., dict$event_global)
                      ~ dict$count_entropic < ctime_cutoff$estimate , rho = 0)

ctime_sdf$p.value <- 1 - pchisq(ctime_sdf$chisq, length(ctime_sdf$n) - 1)
ctime_sdf$statistic <-ctime_sdf$chisq


plot(fit, lty = 1:2,
     ylab = 'event global survival probability',
     xlab = 'time global',
     main = sprintf('%s\ncoxph pvalue %s statistic %s\nlogrank pvalue %s statistic %s',
                    tag, tag2,
                    sprintf('%g', coef(ctime)[5]), sprintf('%g', coef(ctime)[1]),
                    sprintf('%g', ctime_sdf$p.value),  sprintf('%g', ctime_sdf$statistic)))

legend(
    "bottomleft",
    legend = c(sprintf('x < %s', ctime_cutoff$estimate),
               sprintf('x >= %s', ctime_cutoff$estimate)),
    lty = 1:2,
    horiz = FALSE,
    bty = 'n')

plot(ctime_cutoff, xlab = tag, main = 'cut-off selection global survival')
    

```



# Session

```{r sessionInfo, cache = 0}

date()
devtools::session_info()

```
